### PR TITLE
Don't generate patch for a no-change set operation

### PR DESCRIPTION
**Pull request by eriksunsol**
Extending the example in README.md like this:

```JS
var myObj = { firstName:"Joachim", lastName:"Wester", contactDetails: { phoneNumbers: [ { number:"555-123" }] } };
var jsonPatcherProxy = new JSONPatcherProxy( myObj );
var observedObject = jsonPatcherProxy.observe(true);
observedObject.firstName = "Albert";
observedObject.firstName = "Albert";
observedObject.firstName = "Albert";
observedObject.contactDetails.phoneNumbers[0].number = "123";
observedObject.contactDetails.phoneNumbers.push({number:"456"});
var patches = jsonPatcherProxy.generate();
```

results in the following patch:

```JSON
[
  {"op":"replace","path":"/firstName","value":"Albert"},
  {"op":"replace","path":"/firstName","value":"Albert"},
  {"op":"replace","path":"/firstName","value":"Albert"},
  {"op":"replace","path":"/contactDetails/phoneNumbers/0/number","value":"123"},
  {"op":"add","path":"/contactDetails/phoneNumbers/1","value":{"number":"456"}}
]
```

The second and third time `firstName` is set are effectively no-ops which have already been communicated. I think this is wrong. The proposed change recognizes this and renders the following result instead:

```JSON
[
  {"op":"replace","path":"/firstName","value":"Albert"},
  {"op":"replace","path":"/contactDetails/phoneNumbers/0/number","value":"123"},
  {"op":"add","path":"/contactDetails/phoneNumbers/1","value":{"number":"456"}}
]
```

### CODE REVIEW

**Comment by eriksunsol** [comment]
I guess the proposed change doesn't cover no-change set operations on array elements...

File: src/jsonpatcherproxy.js [hunk 1]
```diff
@@ -136,7 +136,11 @@ const JSONPatcherProxy = (function() {
       }
       operation.op = 'add';
       if (tree.hasOwnProperty(key)) {
-        if (typeof tree[key] !== 'undefined' || isTreeAnArray) {
+        if (typeof tree[key] !== 'undefined') {
+          if (tree[key] == newValue)
+            return true; // Value wasn't actually changed, pretend set was successful but don't generate a patch
```
Comment (warpech) [inline src/jsonpatcherproxy.js]: Wrap the `return true` statement in curly braces for readability

Comment (eriksunsol) [reply src/jsonpatcherproxy.js]: Good point. Done!

**Comment by eriksunsol** [comment]
Tests added. I have also worked through no-change operations on array elements and made sure `undefined` and `null` are treated property with respect to their JSON projections in objects and arrays (which is slightly different). I took the liberty to add a test or two which are variants of tests that already existed, but for the special cases e.g. when an `undefined` property is set to `null`. This is to make sure testing for no-change operations do not catch cases which should be left alone.

File: src/jsonpatcherproxy.js [hunk 2]
```diff
@@ -146,6 +152,44 @@ const JSONPatcherProxy = (function() {
...
+   */
+  function isSignificantChangeInArray(oldValue, newValue) {
+    if (typeof oldValue === 'undefined') {
+      oldValue = null;
+    }
+    if (typeof newValue === 'undefined') {
+      newValue = null;
+    }
+    return oldValue !== newValue;
+  }
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: I think this function (and tests) need to be extended cover scenario of adding a `null` beyond the end of the array:
```
arr = Array(5);
JSON.stringify(arr); // "[null,null,null,null,null]"
arr[7]; // undefined
arr[7] = null; // null - Even though `typeof oldValue === 'undefined'`, this is a significant change, as
JSON.stringify(arr); // [null,null,null,null,null,null,null,null]"
```

**Comment by tomalec** [comment]
@warpech suggested to make `return Reflect.set(tree, key, newValue);` DRYier

**Comment by warpech** [comment]
I will make the required changes in a new PR.

**Comment by warpech** [comment]
> @warpech suggested to make return Reflect.set(tree, key, newValue); DRYier

@tomalec I have made two alternative solutions:

1. single `Reflect.set` before the big pile of IFs with many return statements: https://github.com/Palindrom/JSONPatcherProxy/commit/43f3439dd5b43a1d9eb7f348074b3f76c4958e8c
2. single `Reflect.set` after the big pile of IFs with single return statement: https://github.com/Palindrom/JSONPatcherProxy/commit/71078c341ad83d8313c8c13a65178f5f9ba8ece6

The code is quite complex, so I am not very happy with the readability of either. Any suggestions which is better? (I can accept "Current" as an answer, too)

File: src/jsonpatcherproxy.js (hunk 2 above)
Comment (eriksunsol) [reply src/jsonpatcherproxy.js]: @tomalec Good catch! I noticed that these cases are were not dealt with properly and had intended to file a separate issue about it. The thing is that adding `undefined` elements is broken on `master` as well and therefore i figured it was a bit out of scope for this PR.

Check out failing test on 7cf32fd09593fef2751abc51114267e4a14f3897 which is based on current `master`. Here the `undefined` element results in a `replace` patch when it should be an `add`. Based on this PR, the same test also fails, although differently: 5de895f7b922c4222e5a020b08190b67bec904f7. Here the `undefined` element is simply ignored. Note that both fail on elements which is added implicitly, without ever being set such as `arr[5]` and `arr[6]` in the example which are not present in the patch at all.

The test is based on your example with an addition:

```JS
arr = Array(5);
JSON.stringify(arr); // "[null,null,null,null,null]"
arr[7]; // undefined
arr[7] = null;
arr[8] = undefined;
// null - Even though `typeof oldValue === 'undefined'`, these are significant changes, as
JSON.stringify(arr); // [null,null,null,null,null,null,null,null,null]"
```

**Comment by tomalec** [comment]
I have a slight preference over 1. because of readable variable names `wasKeyInTreeBeforeReflection, valueBeforeReflection` and the fact that whether the callback/following code is executed is controlled using standard `return`, not enigmatic `operation.op`

File: src/jsonpatcherproxy.js (hunk 2 above)
Comment (eriksunsol) [reply src/jsonpatcherproxy.js]: See #52.

**Comment by warpech** [comment]
As discussed today between @eriksunsol, @tomalec and me, I have pushed my solution that fixes the test created by @eriksunsol (https://github.com/Palindrom/JSONPatcherProxy/pull/45/commits/0257a923ad761dc0ec9507c869039f57d7142170).

@tomalec, @eriksunsol could you pls check and give your blessing?

I plan to add more tests from https://github.com/Starcounter-Jack/JSON-Patch/tree/master/test in a separate PR before next release.

File: src/jsonpatcherproxy.js (hunk 2 above)
Comment (tomalec) [reply src/jsonpatcherproxy.js]: We found another issue with this function. Is it called when you `pop` an array?
Then change from `undefined`(mapped to `null`) to not existing element.
I guess it should call `deleteTrap` but it's worth to add a test for such case as well.

File: src/jsonpatcherproxy.js [hunk 3]
```diff
@@ -113,25 +113,42 @@ const JSONPatcherProxy = (function() {
       }
     }
     // let's start with this operation, and may or may not update it later
+    const valueBeforeReflection = tree[key];
+    const wasKeyInTreeBeforeReflection = tree.hasOwnProperty(key);
+    if (isTreeAnArray && !isNonSerializableArrayProperty) {
+      const index = parseInt(key, 10);
+      if (index > tree.length) {
+        trapForSet(instance, tree, (index - 1) + '', undefined);
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: Could you add a code comment that this calls set trap for automatically added `undefined`->`null` elements of an array?

Comment (warpech) [reply src/jsonpatcherproxy.js]: Done

**Comment by warpech** [comment]
> We found another issue with this function. Is it called when you `pop` an array?
> Then change from `undefined`(mapped to `null`) to not existing element.
I guess it should call `deleteTrap` but it's worth to add a test for such case as well.

Let's continue that in https://github.com/Palindrom/JSONPatcherProxy/issues/52

**Pull request merged**